openflow15/nxt_message.go

```diff
@@ -1235,33 +1237,60 @@ func (p *PacketIn2) Len() (n uint16) {
 }
 
 func (p *PacketIn2) MarshalBinary() (data []byte, err error) {
-	data = make([]byte, p.Len())
+	return marshalMultipleProperties(p.Props)
+}
+
+func (p *PacketIn2) UnmarshalBinary(data []byte) error {
+	props, err := unMarshalPacketIn2MultipleProperties(data)
+	if err != nil {
+		return err
+	}
+	p.Props = append(p.Props, props...)
+	return nil
+}
+
+func marshalMultipleProperties(props []Property) (data []byte, err error) {
+	totalLen := 0
+	for _, prop := range props {
+		totalLen += getPacketIn2PropertyLength(prop)
+	}
+
+	data = make([]byte, totalLen)
 	n := 0
 
-	for _, prop := range p.Props {
+	for _, prop := range props {
 		b, err := prop.MarshalBinary()
 		if err != nil {
 			return nil, err
 		}
 		copy(data[n:], b)
-		n += int(prop.Len())
+		n += getPacketIn2PropertyLength(prop)
 	}
 	return
 }
 
-func (p *PacketIn2) UnmarshalBinary(data []byte) error {
-	n := 0
+func unMarshalPacketIn2MultipleProperties(data []byte) ([]Property, error) {
+	var props []Property
+	var err error
 
+	n := 0
 	for n < len(data) {
 		prop, err := DecodePacketIn2Prop(data[n:])
 		if err != nil {
 			break
 		}
-		p.Props = append(p.Props, prop)
-		n += int(prop.Len())
+		props = append(props, prop)
+		n += getPacketIn2PropertyLength(prop)
 	}
+	return props, err
+}
 
-	return nil
+func getPacketIn2PropertyLength(prop Property) int {
```

Comment:
The `getPacketIn2PropertyLength` function returns the length of any property, not just `PacketIn2PropPacket`, so I am not sure why the function is named like this.

Reply:
Renamed as `getPropertyLength`.